doc/hardware/cache/guide.rst

```diff
@@ -0,0 +1,183 @@
+
+.. note::
+
+  While cache coherence can be a concern for data shared between SMP cores, Zephyr
+  in general ensures that memory will be seen in a coherent state from multiple
+  cores. Most applications will only need to use the cache APIs for interaction
+  with external hardware like DMA controllers or foreign CPUs running a
+  different OS image. For more information on cache coherence between SMP cores,
+  see :kconfig:option:`CONFIG_KERNEL_COHERENCE`.
+
+When dealing with memory shared between a processor core and other bus masters,
+cache coherency needs to be considered. Typically processor caches exist as
+close to each processor core as possible to maximize performance gain. Because
+of this, data moved into and out of memory by DMA engines will be stale in the
+processor's cache, resulting in what appears to be corrupt data. If you are
+moving data using DMA and the processor doesn't see the data you expect, cache
+coherency may be the issue.
+
+There are multiple approaches to ensuring that the data seen by the processor
+core and peripherals is coherent. The simplest is just to disable caching, but
+this defeats the purpose of having a hardware cache in the first place and
+results in a significant performance hit. Many architectures provide methods for
+disabling caching for only a portion of memory. This can be useful when cache
+coherence is more important than performance, such as when using DMA with SPI.
+Finally, there is the option to flush or invalidate the cache for regions of
+memory at runtime.
+
+Globally Disabling the Data Cache
+---------------------------------
+
+As mentioned above, globally disabling data caching can have a significant
+performance impact but can be useful for debugging.
+
+Requirements:
+
+* :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+
+* :kconfig:option:`CONFIG_CACHE_MANAGEMENT`: cache API enabled.
+
+* Call :c:func:`sys_cache_data_disable()` to globally disable the data cache.
+
+Disabling Caching for a Memory Region
+-------------------------------------
+
+Disabling only a portion of memory can be a good performance compromise if
+performance on the uncached memory is not critical to the application. This is a
+good option if the application requires many small unrelated buffers that are
+smaller than a cache line.
+
+Requirements:
+
+* :kconfig:option:`CONFIG_DCACHE`: DCACHE control enabled in Zephyr.
+
+* :kconfig:option:`CONFIG_MEM_ATTR`: enable the ``mem-attr`` library for
+  handling memory attributes in the device tree.
+
+* Annotate your device tree according to :ref:`mem_mgmt_api`.
+
+Assuming the MPU driver is enabled, it will configure the specified regions
+according to the memory attributes specified during kernel initialization. When
+using a dedicated uncached region of memory, the linker needs to be instructed
+to place buffers into that region. This can be accomplished by specifying the
+memory region explicitly using ``Z_GENERIC_SECTION``:
```

Comment:
Wow, I had no idea that existed. That seems useful but that capability only seems able to relocate both `.text` and `.data`. Relocating buffers into an uncached region using this mechanism is possible, I suppose, but comes along with all the impact of relocating `.text` to that same region. I feel like `zephyr_code_relocate` is more valuable for placing compiled code that you don't want to modify (third-party DSP code, for example) into a special region than it is for mapping buffers to an uncached region. I'd like to leave references to `zephyr_code_relocate` out of this particular document as I think it'll mostly confuse people. If you feel strongly that it should be included, let me know.

Reply:
Actually, reading through the docs it appears that you CAN relocate just data. I'll add a small note as this will enable users to relocate buffers without modifying code, but I think I would only recommend this as a last resort.

Reply:
I keep waffling on this. Thinking more about it, it's also technically possible for users to relocate individual buffers or functions by modifying the linker code, but I wouldn't want to add that to this document. Anybody doing things with that kind of complexity *probably* aren't the target audience for this document. If you still think a reference to `zephyr_code_relocate` should be added to this document, I have the changes typed up and I'll push them, but I'm planning on leaving it out for now.

Reply:
Maybe @npitre or @dcpleung or @carlocaione want to comment here (re: documentation on how to put data in uncahed memory regions).